libs/telemetry/logger.go

```diff
@@ -0,0 +1,141 @@
+package telemetry
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"time"
+
+	"github.com/databricks/cli/libs/log"
+	"github.com/databricks/cli/libs/telemetry/protos"
+	"github.com/databricks/databricks-sdk-go/apierr"
+	"github.com/databricks/databricks-sdk-go/client"
+	"github.com/databricks/databricks-sdk-go/config"
+	"github.com/google/uuid"
+)
+
+// Environment variable to disable telemetry. If this is set to any value, telemetry
+// will be disabled.
+const DisableEnvVar = "DATABRICKS_CLI_DISABLE_TELEMETRY"
+
+func Log(ctx context.Context, event protos.DatabricksCliLog) {
+	fromContext(ctx).log(event)
+}
+
+func SetExecutionContext(ctx context.Context, ec protos.ExecutionContext) {
+	fromContext(ctx).setExecutionContext(ec)
+}
+
+func HasLogs(ctx context.Context) bool {
+	return len(fromContext(ctx).getLogs()) > 0
+}
+
+type logger struct {
+	logs []protos.FrontendLog
+}
+
+func (l *logger) log(event protos.DatabricksCliLog) {
+	if l.logs == nil {
+		l.logs = make([]protos.FrontendLog, 0)
+	}
+	l.logs = append(l.logs, protos.FrontendLog{
+		FrontendLogEventID: uuid.New().String(),
+		Entry: protos.FrontendLogEntry{
+			DatabricksCliLog: event,
+		},
+	})
+}
+
+func (l *logger) getLogs() []protos.FrontendLog {
+	return l.logs
+}
+
+func (l *logger) setExecutionContext(ec protos.ExecutionContext) {
+	for i := range l.logs {
+		l.logs[i].Entry.DatabricksCliLog.ExecutionContext = &ec
+	}
+}
+
+func Upload(ctx context.Context, cfg *config.Config) error {
+	l := fromContext(ctx)
+	if len(l.logs) == 0 {
+		return errors.New("no logs to upload")
+	}
+
+	protoLogs := make([]string, len(l.logs))
+	for i, log := range l.logs {
+		b, err := json.Marshal(log)
+		if err != nil {
+			return fmt.Errorf("failed to marshal log: %s", err)
+		}
+		protoLogs[i] = string(b)
+	}
+
+	apiClient, err := client.New(cfg)
+	if err != nil {
+		return fmt.Errorf("failed to create API client: %w", err)
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, 3*time.Second)
+	defer cancel()
+
+	// Only try uploading logs for a maximum of 3 times.
+	for i := range 3 {
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("uploading telemetry logs timed out: %w", ctx.Err())
+		default:
+			// proceed
+		}
+
+		resp, err := attempt(ctx, apiClient, protoLogs)
```

Comment:
If the command timed out, the error equals `context.DeadlineExceeded`.

Reply:
Added a handler for the deadline exceeded error.